In [14]:
import sys
sys.path.append('../src')
from clean_data import clean_chess
import pandas as pd


In [15]:
df = pd.read_csv('../data/raw/chess_games.csv')

df = clean_chess(df)
print(df.head())

   game_id  rated  turns victory_status winner time_increment       white_id  \
0        1  False     13    Out of Time  White           15+2       bourgris   
1        2   True     16         Resign  Black           5+10           a-00   
2        3   True     61           Mate  White           5+10         ischia   
3        4   True     61           Mate  White           20+0  daniamurashov   
4        5   True     95           Mate  White           30+3      nik221107   

   white_rating      black_id  black_rating  ... opening_code opening_moves  \
0          1500          a-00          1191  ...          D10             5   
1          1322     skinnerua          1261  ...          B00             4   
2          1496          a-00          1500  ...          C20             3   
3          1439  adivanov2009          1454  ...          D02             3   
4          1523  adivanov2009          1469  ...          C41             5   

                         opening_fullname   

In [30]:
non_draws = df[df['winner'] != 'Draw']

higher_rated_won = (
    ((non_draws['rating_diff'] > 0) & (non_draws['winner'] == 'White'))
    |
    ((non_draws['rating_diff'] < 0) & (non_draws['winner'] == 'Black'))
)

higher_rated_won.mean() * 100

np.float64(64.643081431861)

Q7: After adding rating_diff, what % of games did the higher-rated player win?64.643081431861

In [31]:
df['is_suspicious'].sum()

np.int64(342)

Q8: How many games are flagged as suspicious? 342 less than 5 turn

In [32]:
df['opening_family'].nunique()

227

Q9: How many unique opening families exist? 227

In [16]:
df['winner'].value_counts(normalize=True) * 100

winner
White    49.860405
Black    45.403330
Draw      4.736265
Name: proportion, dtype: float64

In [41]:
df['victory_status'].value_counts(normalize=True)*100

victory_status
Resign         55.573836
Mate           31.533553
Out of Time     8.375710
Draw            4.516901
Name: proportion, dtype: float64

Q10 What is the win rate for White, Black, and Draw? (% of total games) white : 49.86/ black : 45.40 / draw : 4.732

In [18]:
df.groupby('victory_status')['turns'].mean()

victory_status
Draw           83.781457
Mate           65.415020
Out of Time    72.742857
Resign         53.912533
Name: turns, dtype: float64

Which victory_status has the highest average number of turns? Draw games has 83.78 turn 

In [46]:
df[df['winner'] == 'Black']['opening_family'].value_counts().head(1)

opening_family
Sicilian Defense    1273
Name: count, dtype: int64

In [47]:
df[df['winner'] == 'White']['opening_family'].value_counts().head(1)

opening_family
Sicilian Defense    1173
Name: count, dtype: int64

Which opening family is most popular when Black wins? Same for White? black : 1273 / white : 1173

In [55]:
(df['winner'] == 'White').groupby(df['rated']).mean() * 100

rated
False    49.935947
True     49.842154
Name: winner, dtype: float64

Q14 Do rated games have a different White win rate than unrated games? white rated: 49.8 / white unrated : 49.9

In [58]:
def classify_game(turns):
    if turns < 20:
        return 'Short'
    elif turns < 50:
        return 'Medium'
    else:
        return 'Long'
df['game_length'] = df['turns'].apply(classify_game)
df['game_length'].value_counts(normalize=True) * 100

game_length
Long      58.246086
Medium    33.472929
Short      8.280985
Name: proportion, dtype: float64

Q15 Classify each game as Short/Medium/Long using apply(). What % is each? long : 58.24 / medium : 33.427 / short : 8.28